crates/uv-toolchain/src/discovery.rs

```diff
@@ -361,69 +361,89 @@ fn python_interpreters<'a>(
     sources: &ToolchainSources,
     cache: &'a Cache,
 ) -> impl Iterator<Item = Result<(ToolchainSource, Interpreter), Error>> + 'a {
-    python_executables(version, implementation, sources)
-        .map(|result| match result {
-            Ok((source, path)) => Interpreter::query(&path, cache)
-                .map(|interpreter| (source, interpreter))
-                .inspect(|(source, interpreter)| {
-                    debug!(
-                        "Found {} {} at `{}` ({source})",
-                        LenientImplementationName::from(interpreter.implementation_name()),
-                        interpreter.python_full_version(),
-                        path.display()
-                    );
-                })
-                .map_err(Error::from)
-                .inspect_err(|err| debug!("{err}")),
-            Err(err) => Err(err),
-        })
-        .filter(move |result| match result {
-            // Filter the returned interpreters to conform to the system request
-            Ok((source, interpreter)) => match (
-                system,
-                // Conda environments are not conformant virtual environments but we should not treat them as system interpreters
-                interpreter.is_virtualenv() || matches!(source, ToolchainSource::CondaPrefix),
-            ) {
-                (SystemPython::Allowed, _) => true,
-                (SystemPython::Explicit, false) => {
-                    if matches!(
-                        source,
-                        ToolchainSource::ProvidedPath | ToolchainSource::ParentInterpreter
-                    ) {
-                        debug!(
-                            "Allowing system Python interpreter at `{}`",
-                            interpreter.sys_executable().display()
-                        );
-                        true
-                    } else {
-                        debug!(
-                            "Ignoring Python interpreter at `{}`: system interpreter not explicit",
-                            interpreter.sys_executable().display()
-                        );
-                        false
-                    }
-                }
-                (SystemPython::Explicit, true) => true,
-                (SystemPython::Disallowed, false) => {
+    // TODO(zanieb): Move filtering to callers
+    filter_by_system_python(
```

Comment:
This refactor wasn't needed for this pull request but we'll do this in the next change to facilitate respecting the system python preferences when reading executable names.

Reply:
TODO addressed in https://github.com/astral-sh/uv/pull/4310